### ISSUE TITLE
apply_classifier() function

### ISSUE THREAD

**Issue by pngmafia** [ultralytics/yolov5#2700]
@glenn-jocher  trained a custom classifier model using yolov5s and 64 as img_size. got accuracy as 94.1 % accuracy.
loaded the classifier model in detect.py using
```modelc = torch.hub.load('ultralytics/yolov5', "yolov5s", pretrained=True, autoshape=False)
        modelc.model = modelc.model[:8]
        m = modelc.model[-1]  # last layer
        ch = m.conv.in_channels if hasattr(m, 'conv') else sum([x.in_channels for x in m.m])  # ch into module
        c = Classify(ch, 220)  # Classify()
        c.i, c.f, c.type = m.i, m.f, 'models.common.Classify'  # index, from, type
        modelc.model[-1] = c
        modelc.to(device)
        modelc.eval()
        modelc.load_state_dict(torch.load("./classifier.pth", map_location=device))
```
 
but the results are very bad. None of the classes are getting classified correctly. Is there something wrong with the apply_classifier function?
even tried to resize to 64 instead of 224. https://github.com/ultralytics/yolov5/blob/9ccfa85249a2409d311bdf2e817f99377e135091/utils/general.py#L574
still no luck. Please help.

**Comment by glenn-jocher** [comment]
@pngmafia hi there thanks for asking about applying a classifier! 

The classifier training branch is not set up to automatically integrate with detect.py second-stage classifier (though that would be a cool idea), so it's likely that the image preprocessing steps are different. This means that you need to look at classifier.py and use the same exact preprocessing steps there when you run inference with the second stage classifier in detect.py. Everything needs to be exactly the same as the testform transforms.
https://github.com/ultralytics/yolov5/blob/058d667dc953405f76e5c03b85cbd7c4eb70d128/classifier.py#L54-L63

**Comment by pngmafia** [comment]
@glenn-jocher I have tried to replicate everything exactly. 
```
def apply_classifier(x, model, img, im0,device):
    # applies a second stage classifier to yolo outputs
    im0 = [im0] if isinstance(im0, np.ndarray) else im0
    for i, d in enumerate(x):  # per image
        if d is not None and len(d):
            d = d.clone()

            # # Rescale boxes from img_size to im0 size
            scale_coords(img.shape[2:], d[:, :4], im0[i].shape)
            
            trainform = T.Compose([T.RandomGrayscale(p=0.01),
                       T.RandomHorizontalFlip(p=0.5),
                       T.RandomAffine(degrees=1, translate=(.2, .2), scale=(1 / 1.5, 1.5),
                                      shear=(-1, 1, -1, 1), fill=(114, 114, 114)),
                       T.Resize([64, 64]),  # very slow
                       T.ToTensor(),
                       T.Normalize((0.5, 0.5, 0.5), (0.25, 0.25, 0.25))])  # PILImage from [0, 1] to [-1, 1]
            testform = T.Compose(trainform.transforms[-3:])
            
            resize = torch.nn.Upsample(size=(64, 64), mode='bilinear',
                                           align_corners=False)  # image resize

            # Classes
            pred_cls1 = d[:, 5].long()
            ims = []
            temp = None
            for j, a in enumerate(d):  # per item
                cutout = im0[i][int(a[1]):int(a[3]), int(a[0]):int(a[2])]
                cv2.imwrite('test%i.jpg' % j, cutout)
                print("cutout type:",type(cutout),cutout.shape)
                pil_img = Image.fromarray(cutout).convert('RGB')
                pil_img = testform(pil_img)
                print("pil_img shape:",pil_img.shape)
                pil_img = pil_img.unsqueeze(0)
                pil_img = resize(pil_img.to(device))
                if temp is None:
                    temp = pil_img
                else:
                    temp = torch.cat([temp, pil_img], dim=0)

            print("comb:",temp.shape)
            pred_cls2 = torch.max(model(temp), 1)[1]
            print(pred_cls1,pred_cls2)
            x[i] = x[i][pred_cls1 == pred_cls2]  # retain matching class detections

    return x
```
But the results are still off. 
example:
pred_cls1 vs pred_cls2:
tensor([5], device='cuda:0') tensor([7], device='cuda:0')
tensor([0, 0], device='cuda:0') tensor([1, 1], device='cuda:0')
tensor([0], device='cuda:0') tensor([1], device='cuda:0')

Though it looks like the results are very close. 
could you please confirm if the updated apply_classifier() function is correct.

**Comment by glenn-jocher** [comment]
@pngmafia sorry I can't comment on custom code, but if you are following my earlier guidance you should be on the right track!
